### ISSUE TITLE
Bundler 2.2+ looks for gem dependencies in Gemfury when https://rubygems.org primary source

### ISSUE THREAD

**Issue by patbl** [gemfury/gemfury#91]
**TL;DR: Does Gemfury work with Bundler 2.2+? It seems not to for me.**

I realize this isn't related to the Gemfury CLI, but I wanted to post somewhere public where this issue might get attention (I hit a wall with customer support).

For me, Gemfury works fine with Bundler 2.1. But in 2.2 and later versions, for gems sourced from Gemfury, it attempts to get these gems' dependencies from Gemfury (even when rubygems.org is specified as the primary source).

Here's a reproduction:

```
#/usr/bin/env bash

ruby -v
gem -v
bundle -v

mkdir bundler-test
cd bundler-test
wget https://rubygems.org/downloads/rspec-rails-6.0.1.gem
fury push rspec-rails-6.0.1.gem
cat <<EOF> Gemfile
source "https://rubygems.org/"
gem "rspec-rails", source: "https://[redacted]@gem.fury.io/[redacted]/"
EOF
bundle install --verbose
```

This downloads `rspec-rails` (a random gem that has some dependencies), pushes it to Gemfury, creates a Gemfile that sources `rspec-rails` from Gemfury and has rubygems.org as a primary source, and runs `bundle install`.

Here's the output:

```
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-darwin20]
3.4.10
Bundler version 2.4.10
mkdir: bundler-test: File exists
--2023-05-03 11:56:23--  https://rubygems.org/downloads/rspec-rails-6.0.1.gem
Resolving rubygems.org (rubygems.org)... 151.101.1.227, 151.101.65.227, 151.101.193.227, ...
Connecting to rubygems.org (rubygems.org)|151.101.1.227|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 70144 (68K) [binary/octet-stream]
Saving to: ‘rspec-rails-6.0.1.gem’

rspec-rails-6.0.1.gem                                      100%[=======================================================================================================================================>]  68.50K  --.-KB/s    in 0.02s

2023-05-03 11:56:23 (3.71 MB/s) - ‘rspec-rails-6.0.1.gem’ saved [70144/70144]

Top level ::CompositeIO is deprecated, require 'multipart/post' and use `Multipart::Post::CompositeReadIO` instead!
Top level ::Parts is deprecated, require 'multipart/post' and use `Multipart::Post::Parts` instead!
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
Uploading rspec-rails-6.0.1.gem - done
Running `bundle install --verbose` with bundler 2.4.10
Found changes from the lockfile, re-resolving dependencies because the dependencies in your gemfile changed, you added a new platform to your gemfile
HTTP GET https://gem.fury.io/[redacted]/me/versions
HTTP 304 Not Modified https://gem.fury.io/[redacted]/me/versions
Fetching gem metadata from https://gem.fury.io/[redacted]/me/
Looking up gems ["rspec-rails"]
Bundler::GemNotFound: Could not find gem 'rspec-rails' in rubygems repository https://gem.fury.io/[redacted]/me/.

/Users/pat/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/bundler/resolver.rb:290:in `raise_not_found!'
  /Users/pat/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/bundler/resolver.rb:343:in `block in prepare_dependencies'
  [snip]
  /Users/pat/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/bundler/friendly_errors.rb:117:in `with_friendly_errors'
  /Users/pat/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.4.10/exe/bundle:33:in `<top (required)>'
  /Users/pat/.asdf/installs/ruby/3.2.2/bin/bundle:25:in `load'
  /Users/pat/.asdf/installs/ruby/3.2.2/bin/bundle:25:in `<main>'
```

I wonder whether [this issue](https://github.com/rubygems/rubygems/issues/4694) is relevant.

**Comment by pboling** [comment]
Oh fuck.  About to cancel GemFury subscription.  It seems like they don't even have any engineers working there.

**Comment by rykov** [comment]
@patbl We are still having trouble reproducing this, and we have not received similar problem reports from others. Here is what I see on a fresh install of Ruby 3.2.2 on macOS with RBEnv:

```
$ ls -a; ruby -v; gem -v; bundle -v; uname -a
. ..  Gemfile
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [arm64-darwin22]
3.4.10
Bundler version 2.4.10
Darwin [redacted] 22.5.0 Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000 arm64
$
$ more Gemfile
source "https://rubygems.org/"
gem "rspec-rails", source: "https://[redacted]@gem.fury.io/me/"
$
$ bundle install --verbose --path vendor/bundle
Running `bundle install --path "vendor/bundle" --verbose` with bundler 2.4.10
[DEPRECATED] The `--path` flag is deprecated because it relies on being remembered across bundler invocations, which bundler will no longer do in future versions. Instead please use `bundle config set --local path 'vendor/bundle'`, and stop using this flag
Found changes from the lockfile, re-resolving dependencies because the dependencies in your gemfile changed, you added a new platform to your gemfile
HTTP GET https://[redacted]@gem.fury.io/me/versions
HTTP 200 OK https://[redacted]@gem.fury.io/me/versions
Fetching gem metadata from https://gem.fury.io/me/
Looking up gems ["rspec-rails"]
HTTP GET https://[redacted]@gem.fury.io/me/info/rspec-rails
HTTP 200 OK https://[redacted]@gem.fury.io/me/info/rspec-rails
Looking up gems ["actionpack", "activesupport", "railties", "rspec-core", "rspec-expectations", "rspec-mocks", "rspec-support"]
Resolving dependencies...
HTTP GET https://index.rubygems.org/versions
HTTP 200 OK https://index.rubygems.org/versions
Fetching gem metadata from https://rubygems.org/
Looking up gems [...snip]
Fetching rake 13.0.6
Installing rake 13.0.6
[... snip]
Bundle complete! 1 Gemfile dependency, 29 gems now installed.
Bundled gems are installed into `./vendor/bundle`
$ 
```

My two suspicions would be a possible Bundler incompatibility (a plugin or a modified dependency) that alters its behavior, or, more likely, it is a caching issue. Here are my recommendations and things to try:

- Remove `~/.bundler/cache` directory before installing.
- Try this on a clean installation of Ruby or in a Docker container.
- Never upload a different file with the same version (i.e. delete/replace method).
- Update to the latest RubyGems and Bundler.

**Comment by rykov** [comment]
@pboling This one is on us. We haven't been clear about the deprecation of our [Ruby-based CLI](https://github.com/gemfury/gemfury) in favor of the [new CLI](https://github.com/gemfury/cli). This repository will become the Ruby SDK, however, we haven't prioritized that work yet. In the meantime, we'll take care of the Faraday dependency.
